File: frontend/src/services/administration/userService.ts
import { api, apiService } from '../core/apiService';
import { User } from './types';
import { PageResponse } from '../core/types';
import { API_URL } from '../../config/apiConfig';

interface CurrentUser {
    id: number;
    username: string;
    email: string;
    roles: string[];
}

export const userService = {
    getCurrentUser: async (): Promise<CurrentUser> => {
        const response = await api.get<CurrentUser>(`${API_URL}/api/user`);
        return response.data;
    },

    getAllUsers: async (page: number = 0, size: number = 10, sortBy: string = 'username', direction: string = 'asc'): Promise<PageResponse<User>> => {
        const response = await api.get<PageResponse<User>>(`${API_URL}/api/users`, {
            params: { page, size, sortBy, direction }
        });
        return response.data;
    },

    searchUsers: async (searchTerm: string, page: number = 0, size: number = 10): Promise<PageResponse<User>> => {
        const response = await api.get<PageResponse<User>>(`${API_URL}/api/users`, {
            params: { searchTerm, page, size }
        });
        return response.data;
    },

    getUsersByStatus: async (enabled: boolean, page: number = 0, size: number = 10): Promise<PageResponse<User>> => {
        const response = await api.get<PageResponse<User>>(`${API_URL}/api/users`, {
            params: { enabled, page, size }
        });
        return response.data;
    },

    getLockedUsers: async (page: number = 0, size: number = 10, sortBy: string = 'username', direction: string = 'asc'): Promise<PageResponse<User>> => {
        const response = await api.get<PageResponse<User>>(`${API_URL}/api/users/locked`, {
            params: { page, size, sortBy, direction }
        });
        return response.data;
    },

    createUser: async (user: Omit<User, 'id'>): Promise<User> => {
        console.log('Creating user with data:', { ...user, password: '***' });
        const response = await api.post<User>(`${API_URL}/api/users`, user);
        return response.data;
    },

    updateUser: async (id: number, user: Partial<User>): Promise<User> => {
        const response = await api.put<User>(`${API_URL}/api/users/${id}`, user);
        return response.data;
    },

    deleteUser: async (id: number): Promise<void> => {
        await api.delete(`${API_URL}/api/users/${id}`);
    },

    changePassword: async (id: number, oldPassword: string, newPassword: string): Promise<void> => {
        await api.post(`${API_URL}/api/users/${id}/change-password`, null, {
            params: { oldPassword, newPassword }
        });
    },

    resetPassword: async (email: string): Promise<void> => {
        await api.post(`${API_URL}/api/users/reset-password`, null, {
            params: { email }
        });
    },

    unlockAccount: async (id: number): Promise<void> => {
        await api.post(`${API_URL}/api/users/${id}/unlock`);
    }
};
   

File: backend/processor/src/main/java/com/middleware/processor/service/interfaces/UserService.java
package com.middleware.processor.service.interfaces;

import com.middleware.shared.model.User;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.core.userdetails.UserDetailsService;
import java.util.Optional;

/**
 * Service for managing User entities.
 */
public interface UserService extends UserDetailsService {
    /**
     * Get all users with pagination.
     *
     * @param pageable The pagination information
     * @return Page of users
     */
    Page<User> getAllUsers(Pageable pageable);

    /**
     * Get a user by ID.
     *
     * @param id The ID of the user
     * @return Optional containing the user if found
     */
    Optional<User> getUserById(Long id);

    /**
     * Get a user by username.
     *
     * @param username The username
     * @return Optional containing the user if found
     */
    Optional<User> getUserByUsername(String username);

    /**
     * Create a new user.
     *
     * @param user The user to create
     * @return The created user
     */
    User createUser(User user);

    /**
     * Update an existing user.
     *
     * @param id The ID of the user to update
     * @param user The updated user data
     * @return The updated user
     */
    Optional<User> updateUser(Long id, User user);

    /**
     * Delete a user.
     *
     * @param id The ID of the user to delete
     */
    void deleteUser(Long id);

    /**
     * Change a user's password.
     *
     * @param id The ID of the user
     * @param oldPassword The old password
     * @param newPassword The new password
     * @return true if password was changed successfully
     */
    boolean changePassword(Long id, String oldPassword, String newPassword);

    /**
     * Reset a user's password.
     *
     * @param id The ID of the user
     * @param newPassword The new password
     */
    void resetPassword(Long id, String newPassword);

    /**
     * Check if a username is available.
     *
     * @param username The username to check
     * @return true if username is available
     */
    boolean isUsernameAvailable(String username);

    Page<User> getUsers(Pageable pageable, String searchTerm, Boolean enabled);
} 
